kafka-4576: Log segments close to max size break on fetch

huxihx opened the pull request:
JVM spec does not make a guarantee that the buffer will be filled up when invoking FileChannel.read, which is called in searchOffsetWithSize to fail the follower replicas' reading data from leader when large log segment is set.

huxihx committed:
kafka-4576: Log segments close to max size break on fetch

JVM spec does not make a guarantee that the buffer will be filled up when invoking FileChannel.read, which is called in searchOffsetWithSize to fail the follower replicas' reading data from leader when large log segment is set.

huxihx commented:
@ijuma @onurkaraman Please have a review on this PR. Thanks

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -784,4 +784,39 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+     * @throws IllegalArgumentException If invalid arguments were passed
+     * @throws IOException If some other I/O error occurs
+     */
+    public static void readFullyFrom(FileChannel channel, ByteBuffer buffer, long startPosition) throws IOException {
```
Maybe the name should simply be `readFully`. And `buffer` should maybe be `destinationBuffer`. And `startPosition` can maybe just be `position` for consistency with `FileChannel.read`.

Finally, we need a mechanism to indicate that EOF has been reached. Maybe a `boolean` is good enough. The usual way of returning an `int` is a bit weird for the case where some bytes are read and then EOF is reached.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -784,4 +784,39 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+        }
+        if (buffer == null) {
+            throw new IllegalArgumentException("Must specify the byte buffer for the channel.");
+        }
```
I don't think we need the `null` checks.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -784,4 +784,39 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+            throw new IllegalArgumentException("Must specify the byte buffer for the channel.");
+        }
+        if (startPosition < 0) {
+            throw new IllegalArgumentException("The file position cannot be negative.");
```
We should include `startPosition` in the message.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -784,4 +784,39 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+        if (startPosition < 0) {
+            throw new IllegalArgumentException("The file position cannot be negative.");
+        }
+        final int totalBytesToRead = buffer.capacity();
```
This is not correct, it should be `buffer.remaining`. It seems simpler to just do `while (buffer.remaining())` with an early exit in the case read returns `-1`.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -784,4 +784,39 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+            int bytesRead = channel.read(buffer, currentPosition);
+            if (bytesRead == -1) {
+                log.warn(String.format("Read incomplete data from the given file channel. Current file channel size: %d, and wanted to read at position: %d",
+                        channel.size(), currentPosition));
```
What's the purpose of this warning? It doesn't seem needed.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -784,4 +784,39 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+            totalBytesRead += bytesRead;
+            currentPosition += bytesRead;
+        }
+        buffer.flip(); // Make the buffer ready for read
```
A bit unclear why we need this. In my mind, `readFully` should be as close to `FileChannel.read` as possible with the exception that it attempts to fill the buffer while end of file is not reached.

ijuma reviewed (changes requested):
Thanks for the PR. In addition to the other review comments:
1. It would be good to add unit tests for the new method
2. FileRecords.readInto also has to be updated.

cc @junrao @hachikuji who may have thoughts on this as well.

ijuma commented:
Note that there are a number of test failures caused by this change.

huxihx commented:
Thanks for the comments. Will work on these comments asap.

huxihx committed:
kafka-4576: Log segments close to max size break on fetch

Addressed Ijuma's comments
1. Refined some variaiable names
2. Fix code bug to have all unit test cases passed

huxihx commented:
@ijuma Thanks for the comments. Refined the code as per your suggestions. Now all the unit test cases should pass. Since we already have about 7 unit test cases (such as FileRecordsTest.testRead), do you think we still need to add new test cases for this method? Thanks again.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -784,4 +784,29 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+    public static void readFully(FileChannel channel, ByteBuffer destinationBuffer, long position) throws IOException {
+        if (channel == null) {
+            throw new IllegalArgumentException("Cannot read an empty file channel.");
+        }
```
I don't think we need this `null` check either.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -784,4 +784,29 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
     }
 
+    /**
+     * Read data from the channel to the given byte buffer
```
Maybe add at the end `until there are no bytes remaining in the buffer or the channel has reached end of stream`.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -784,4 +784,29 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+     * @param position The file position at which the transfer is to begin; must be non-negative
+     *
+     * @throws IllegalArgumentException If invalid arguments were passed
+     * @throws IOException If some other I/O error occurs
```
@throws IOException If an I/O error occurs. See {@link FileChannel#read(ByteBuffer, long)} for detailed information on the exception cases.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -784,4 +784,29 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+            throw new IllegalArgumentException("Cannot read an empty file channel.");
+        }
+        if (position < 0) {
+            throw new IllegalArgumentException("The file position cannot be negative, but we got " + position);
```
Nit: "The file channel position..."

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/FileRecords.java:
```diff
@@ -113,7 +113,7 @@ public FileChannel channel() {
      * @throws IOException
      */
     public ByteBuffer readInto(ByteBuffer buffer, int position) throws IOException {
```
Worth clarifying the docs, similar to the `readFully` documentation.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -784,4 +784,29 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+            throw new IllegalArgumentException("The file position cannot be negative, but we got " + position);
+        }
+        long currentPosition = position;
+        while (destinationBuffer.hasRemaining()) {
```
I think we probably want a `do/while` loop here. There should be no difference in behaviour, but it seems to model the problem better (i.e. we first do a read and then we check if there is still space remaining in the buffer. Maybe:

```java
        long currentPosition = position;
        int bytesRead;
        do {
            bytesRead = channel.read(destinationBuffer, currentPosition);
            currentPosition += bytesRead;
        } while (bytesRead != -1 && destinationBuffer.hasRemaining());
```

ijuma reviewed (commented):
Thanks for the updates. Looks good overall, just a few nits.

huxi-2b committed:
Merge branch 'kafka4576_FileChannel_read' of https://github.com/amethystic/kafka into kafka4576_FileChannel_read

huxi-2b committed:
kafka-4576: Log segments close to max size break on fetch

adjusted some comments as per Ijuma's suggestion.

huxi-2b committed:
Merge remote-tracking branch 'upstream/trunk' into kafka4576_FileChannel_read

huxihx commented:
@ijuma already addressed your comments, please take some time to review again. Appreciated.

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/FileRecords.java:
```diff
@@ -108,14 +108,17 @@ public FileChannel channel() {
-     * @return The same buffer
-     * @throws IOException
+     * Read data from the channel to the given byte buffer until there are no bytes remaining in the buffer or
+     * the channel has reached end of stream
```
Hmm, the idea was to extend the existing definition to explain in more detail what happens, not to change to be the same as `readFully`.

huxihx replied:
sorry, but I do not catch your meaning. Do you mean we should modify the comments for readInto method to explain why we use readFully instead of naive Channel.read?

ijuma replied:
The previous comment said "Read log entries into the given buffer". This is more specific than "Read data from the channel to the given byte buffer..." and we should keep that. We could perhaps say:

"Read log entries into the given buffer until there are no bytes remaining in the buffer or the channel has reached end of stream."

huxihx replied:
Thank you for clarifying this. Already addressed. Please check.

huxi-2b committed:
kafka-4576: Log segments close to max size break on fetch

Restored the comments for FileRecords.readInto method and tweaked some
expression.

ijuma reviewed (approved):
LGTM. Will wait for @hachikuji or @junrao to take a look as well before merging.

hachikuji commented:
The fix LGTM. Would be nice to have some test cases for `readFully`. Maybe that wouldn't be too bad using a mock for `FileChannel`?

ijuma commented:
Good point about the test, I had asked for it and then forgot to reply to @amethystic's question. I agree with @hachikuji that it would be useful to have a simple test with a mocked `FileChannel`.

huxihx commented:
@ijuma sorry but not familiar with the mock things. Could you give an example on how to get started with that?

ijuma commented:
@amethystic, if you search for `EasyMock` in our tests, you'll find many examples. The idea is to mock the `FileChannel.read` method to verify that `readFully` works correctly for the various cases: end of stream is reached before buffer is full, buffer is filled with a single read, buffer is filled after multiple reads, etc.

The existing unit tests don't exercise these different paths (since they already passed before your change).

junrao commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -781,4 +781,25 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+        do {
+            bytesRead = channel.read(destinationBuffer, currentPosition);
+            currentPosition += bytesRead;
+        } while (bytesRead != -1 && destinationBuffer.hasRemaining());
```
If bytesRead returns -1, should we throw an IOException?

ijuma replied:
@junrao, that's an interesting suggestion. If we do that, various `if (buf.hasRemaining())` checks in some of the callers no longer make sense.

huxihx replied:
Returning -1 does not mean an abnormal situation. We'd better keep semantics FileChannel.read offers unchanged.

ijuma replied:
@amethystic I believe Jun is suggesting that it is abnormal for `readFully` (i.e. if you can't fill the buffer, then something is wrong). I think a case can be made for that. I think the downside is that the error messages may not be as good as if the callers do the check themselves. The upside is that we avoid the situation where the caller forgets to check.

We'd have to verify that these semantics are right for `FileRecords.readInto` since it doesn't perform any checking atm.

hachikuji replied:
+1 on Jun's suggestion if it works. If not, it might make sense at least to return a flag indicating whether the read completed.

junrao replied:
Yes, it seems that -1 means that we are trying to read beyond the end of the channel, which probably should be treated as abnormal.

huxihx replied:
I think I got what you all mean. Although FileLogInputStream checks whether the buffer has really been filled up after every readFully call, it might be possible that other calling code forgets to do this check (i.e. FileRecords.readInto), so we'd better have the readFully return something that can specify if it's successful to read, no matter throwing an exception or returning a boolean flag. Am I correct? 
If yes, I prefer to return a boolean flag to indicating of the state. What do you think?

ijuma replied:
The boolean makes it easier to do the check, but doesn't help with cases where the caller forgets. So the exception is better from that perspective.

junrao reviewed (commented):
Thanks for the patch. Just a minor comment.

huxihx committed:
refined test case to test EOF for readFully

huxihx commented:
Had readFully throw IOException when reaching EOF and refined test case to test all situations when using readFully. Please review. @ijuma @junrao @hachikuji

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -158,6 +161,39 @@ public void testCloseAll() {
 
+    @Test
+    public void testReadFully() throws IOException {
+        RandomAccessFile raf = new RandomAccessFile(TestUtils.tempFile(), "rw");
```
The issue with using a real file is that you can't test the scenario that we are trying to fix: `FileChannel.read` returns before the buffer is full. See `FileRecordsTest.testTruncateNotCalledIfSizeIsBiggerThanTargetSize` for an example of a mocked fine channel.

huxihx replied:
I can use EasyMock to test throwing IOExceptions for EOF, how can I use it to test "readFully does not return until buffer if full"? A mocked ByteBuffer? If using a mocked bytebuffer, I would have to mock 'hasRemaining' which is a final method. Seems EasyMock does not support to mock final methods. Am I correct?

ijuma replied:
I don't see why you need a mocked ByteBuffer, you can just mock FileChannel read to write less than the buffer remaining?

huxihx replied:
Correct me if I am wrong. Remember readFully has two predicates that are used to check if we can exit the loop: 1. FileChannel.read returns -1; 2. buffer.hasRemaining = false.

I already wrote a unit test case named 'testReadFullyThrowIOExceptionForEOF' to verify readFully can throw IOException when encountering EOF. That covered the first predicate.

Let's see the second predicate: buffer.hasRemaining(). If covering this check, seems I have to use a mocked byte buffer. In doing so can I mock the hasRemaining's return value. Otherwise, a real byte buffer always has remaining since a mocked file channel does no operation to the buffer. 

Do I make myself clear?

ijuma replied:
Why is it the case that the mocked file channel does no operation to the buffer? If you call `readFully` and you have a mocked `FileChannel.read` method, it can (and probably should) add bytes to the buffer. That way you don't need to mock the byte buffer and the test is more realistic.

huxihx replied:
If I use the code below to verify, the position does not change after channelMock.read. Where did I go wrong? Please kindly advice. Thanks.
`

        FileChannel channelMock = EasyMock.createMock(FileChannel.class);
        ByteBuffer byteBuffer = ByteBuffer.wrap("hello, world".getBytes());
        ByteBuffer buffer = ByteBuffer.allocate(12);
        EasyMock.expect(channelMock.write(byteBuffer)).andReturn(12);
        EasyMock.expect(channelMock.read(buffer, 0)).andReturn(12);
        EasyMock.replay(channelMock);
        channelMock.write(byteBuffer);
        channelMock.read(buffer, 0);
        System.out.println(buffer.position());  // always 0 :-(
        EasyMock.verify(channelMock);`

huxihx replied:
@ijuma Seems EasyMock cannot mock functions with side effects. Do you think it's okay I just use EasyMock to test throwing EOF and real files to test multiple reads?

ijuma replied:
Can you not use `andAnswer` instead of `andReturn`?

huxihx replied:
Thanks for the advice. Already refined the test code. Please have a look at it. Thanks.

huxihx commented:
@ijuma Did you get any chances to read my last comments?

huxihx committed:
kafka-4576: Log segments close to max size break on fetch

Addressed Ijuma's comments including:
1. added testReadFullyWithMultiReads with EasyMock to test Utils.readFully could be reading multiple reads filling up the buffer.
2. added testReadFullyThrowIOExceptionForEOF with EasyMock to test Utils.readFully could be throwing IOException when encountering EOF.
3. refined some imports to pass the style check.

huxihx commented:
@ijuma already addressed the comments. Please review again. @hachikuji @junrao  Please also take some time to review. Thank you all.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java:
```diff
@@ -56,7 +57,7 @@ public FileChannelLogEntry nextEntry() throws IOException {
 
         logHeaderBuffer.rewind();
-        channel.read(logHeaderBuffer, position);
+        Utils.readFully(channel, logHeaderBuffer, position);
```
As Ismael said, with the patch, it seems that there is no need to do the logHeaderBuffer.hasRemaining() check in the next line? Ditto in a few other places.

huxihx replied:
Agreed. Already removed those unnecessary checking code. Please review again. Thanks.

junrao reviewed (commented):
@amethystic : Thanks for the patch. Just one more comment.

huxihx committed:
kafka-4576: Log segments close to max size break on fetch

As per Jun's comments, removed all unnecessary code where checks buffer's remaining

junrao commented:
retest this please

amethystic commented:
@junrao clients:test already passed before my committing the fix. Is that what you mean?

huxihx commented:
Seems dozens of tests got failed due to invoking readFully because that -1 is deemed as an exception. There are some cases when we allocate a bigger buffer but cannot fill up it. Should we throw IOException when channel.read returns -1? If yes, all the enclosing methods should be capture this exception and check if this is as designed or just rethrow it.

Do you all think it's okay if I remove the throwing statement and  restore all the outer checking code (check hasRemaining)? Does this make sense?  @junrao @ijuma @hachikuji

huxihx commented:
Throws IOException is only reasonable when we really want to fill up the buffer. But in those cases where buffer might not be filled up (at least for those failed cases), invoking readFully should not throw IOException since returnning -1 is definitely as expected.

huxihx commented:
What readFully must guarantee is that it never leads to a situation where channel still has something to read but buffer does not get filled up.  As a utility method, readFully should not assume channel's EOF is an exceptional case. The enclosing method could take over that responsibility.

junrao commented:
@amethystic : I think this issue is in FileRecords.readInto(). In this method, the expectation is that the buffer may not be filled if there is not enough bytes in the channel. So, for that, we probably need a slightly different util function that reads until the min(Buffer capacity and the remaining bytes in channel). In other places, we do expect the buffer will be filled up.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -781,4 +781,29 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+            currentPosition += bytesRead;
+        } while (bytesRead != -1 && destinationBuffer.hasRemaining());
+        if (bytesRead == -1) {
+            throw new IOException(String.format("Reached EOF before filling up the buffer. Current file channel position = %d, buffer remaining = %d",
```
It would be nice if the calling code could pass a string with context for what is being read. That way we don't lose information like:

```java
if (buf.hasRemaining())
 -                    throw new KafkaException("Failed to read magic byte from FileChannel " + channel);
```

For the case above, we could pass "magic byte" and then the error message would read: "Reached EOF before filling the buffer with magic byte..." or something along those lines.

huxihx replied:
Thanks for the feedback. Personally, I agree @junrao 's suggestion that we might need another read function to take care of the cases when reading EOF is expected. As for @ijuma 's "magic byte", I am not sure if I fully understand how it could be used, but I still think we have put too much responsibilities to readFully which otherwise should do only one thing: READING CHANNEL TO THE BUFFER. So I strongly recommend we refine the enclosing readInto to have all of these things pass the tests. 

What do you think?

ijuma replied:
I am not sure what is your concrete suggestion. Our aim here is to provide internal utility methods that do the right thing concisely and with good error messages.

I agree that it may make sense to have two methods. One could be `readFullyOrFail` and the other could simply be `readFully` maybe. The former would throw the exception on EOF while the other would not.

To clarify my suggestion, I meant that the `readFullyOrFail` variant could take a string parameter that would be included in the IOException message. It's useful to know that we failed to read "magic string", for example.

huxihx replied:
Okay, I will consider to combine all the suggestion and submit a fix asap. Thank for the comments.

huxihx committed:
kafka-4576: Log segments close to max size break on fetch
Addressed Jun's comments
1. Renamed readFully to readFullyOrFail
2. Added a new method 'readFully' that do not throw IOException when encountering EOF
3. Added two unit test cases to test readFully
3. Added two unit test cases to test readFully
3. Added two unit test cases to test readFully

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java:
```diff
@@ -56,9 +57,7 @@ public FileChannelLogEntry nextEntry() throws IOException {
-        channel.read(logHeaderBuffer, position);
-        if (logHeaderBuffer.hasRemaining())
-            return null;
+        Utils.readFullyOrFail(channel, logHeaderBuffer, position);
```
@junrao, is it OK to throw an exception instead of returning `null` if we reach EOF before filling the buffer here?

ijuma committed:
Implementation and documentation improvements

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -158,6 +166,147 @@ public void testCloseAll() {
+            // Scenario 1: test single read
+            Utils.readFullyOrFail(channel, perfectBuffer, 0);
+            assertTrue("Buffer should be filled up.", !perfectBuffer.hasRemaining());
+            // Scenario 2: test multiple reads
```
I am not sure what you mean by `test multiple reads` here. I think we're testing two things here:

1. That the first read doesn't cause a side-effect to the channel that could prevent the second read from succeeding.
2. That we can read into a smaller buffer than what's in the file channel.

"multiple reads" is a little unclear as it sounds like we are causing the underlying file channel to do multiple reads, which I don't think we are.

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -158,6 +166,147 @@ public void testCloseAll() {
+    }
+
+    @Test
+    public void testReadFullyOrFailThrowIOExceptionForEOF() throws IOException {
```
Does this add anything over `testReadFullyOrFailWithRealFile`? If not, we can simply remove it.

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -158,6 +166,147 @@ public void testCloseAll() {
+            assertTrue("Buffer should be filled up.", !perfectBuffer.hasRemaining());
+            // Scenario 2: test multiple reads
+            Utils.readFullyOrFail(channel, smallBuffer, 0);
+            assertTrue("Buffer should be filled up.", !smallBuffer.hasRemaining());
```
Maybe we should add one case where `position > 0`.

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -158,6 +166,147 @@ public void testCloseAll() {
+
+        EasyMock.replay(channelMock);
+        Utils.readFully(channelMock, buffer, 0L);
+        Assert.assertTrue("The buffer should be filled up", !buffer.hasRemaining());
```
Seems like `assertFalse` would be more appropriate here. There are a few cases like that. Also, it would be good to verify the buffer contents.

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -158,6 +166,147 @@ public void testCloseAll() {
     }
 
+    @Test
+    public void testReadFullyOrFailWithRealFile() throws IOException {
```
It would be good to verify that the the buffer is populated correctly.

amethystic replied:
The buffers in this method are all populated by ByteBuffer.allocate. Do you mean we should verify   the semantics for ByteBuffer.allocate?

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -158,6 +166,147 @@ public void testCloseAll() {
+
+        EasyMock.replay(channelMock);
+        Utils.readFullyOrFail(channelMock, buffer, 0L);
+        Assert.assertTrue("The buffer should be filled up", !buffer.hasRemaining());
```
It would be good to verify that the buffer contents are correct as well.

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -158,6 +166,147 @@ public void testCloseAll() {
+                }
+            });
+            totalMockedBytesRead += mockedBytesRead;
+        }
```
Seems like this is the same in `testReadFullyOrFailWithMultiReads`. Maybe we can extract it to a helper method.

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -158,6 +166,147 @@ public void testCloseAll() {
+    }
+
+    @Test
+    public void testReadFullyCanFillUpBuffer() throws IOException {
```
Is this name correct? It seems like the buffer is never filled in this test.

ijuma reviewed (commented):
Thanks for the updates and for the tests. I submitted a PR to your branch with some improvements here:

https://github.com/amethystic/kafka/pull/1

I also left a few comments regarding the tests and one question for Jun regarding the implementation.

huxi-2b committed:
Merge pull request #1 from ijuma/kafka-4576-read-fully

Implementation and documentation improvements

huxihx committed:
kafka-4576: Log segments close to max size break on fetch

Merged Ijuma's fix and addressed Ijuma's comments.

huxihx commented:
@ijuma already merged your fix and addressed the comments. Thanks for the feedback and review again:)

junrao commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -158,6 +167,130 @@ public void testCloseAll() {
+                break;
+            }
+            // If the remaining byte number is less than a step,
+            // directly assign the last mocked bytes read to fix `bufferSize` exactly
```
fix => fit?

junrao commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -781,4 +782,55 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
+     * @param channel File channel containing the data to read from
+     * @param destinationBuffer The buffer into which bytes are to be transferred
+     * @param position The file position at which the transfer is to begin; it must be non-negative
+     * @param description A description of what is being read, this will be included in the EOFException if we
```
Incomplete sentence after "if we"?

junrao reviewed (commented):
@amethystic : Thanks for the patch. LGTM. Just a couple of minor comments.

huxihx committed:
kafka-4576: Log segments close to max size break on fetch
Fixed some typos

huxihx commented:
Sorry for the typos. Thanks.

ijuma commented:
Thanks for the updates, I merged to trunk and 0.10.2 after tweaking the tests a little.
